californium-core/src/test/java/org/eclipse/californium/core/test/LinkFormatTest.java

```diff
@@ -0,0 +1,334 @@
+		serializeTreeOldMatches(root, request2.getOptions().getUriQuery(), buffer);
+		System.out.println("Result for request: /.well-known/core?if=sensor&rt=light-lux  with old matches() method:");
+		System.out.println(buffer.toString() + "\n");
+		Assert.assertEquals(
+				"</sensors/light>;if=\"sensor\";rt=\"light-lux\",",
+				buffer.toString());
+	}
+	
+	@Test
+	public void testMatchesFixed3() {
+		StringBuilder buffer = new StringBuilder();
+		serializeTreeNewMatches(root, request3.getOptions().getUriQuery(), buffer);
+		System.out.println("Result for request: /.well-known/core?if=sensor&rt=angle  with fixed matches() method:");
+		System.out.println(buffer.toString() + "\n");
+		Assert.assertEquals(
+				"",
+				buffer.toString());
+	}
+	
+	@Test
+	public void testMatchesOld3() {
+		StringBuilder buffer = new StringBuilder();
+		serializeTreeOldMatches(root, request3.getOptions().getUriQuery(), buffer);
+		System.out.println("Result for request: /.well-known/core?if=sensor&rt=angle  with old matches() method:");
+		System.out.println(buffer.toString() + "\n");
+		Assert.assertEquals(
+				"",
+				buffer.toString());
+	}
+
+	/**
+	 * Serialize method from LinkFormat class with fixed matches method
+	 * 
+	 * @param resource
+	 * @param queries
+	 * @param buffer
+	 */
+	public static void serializeTreeNewMatches(Resource resource, List<String> queries, StringBuilder buffer) {
+		// add the current resource to the buffer
+		if (resource.isVisible() && matchesFixed(resource, queries)) {
+			buffer.append(LinkFormat.serializeResource(resource));
+		}
+
+		// sort by resource name
+		List<Resource> childs = new ArrayList<Resource>(resource.getChildren());
+		Collections.sort(childs, new Comparator<Resource>() {
+			@Override
+			public int compare(Resource o1, Resource o2) {
+				return o1.getName().compareTo(o2.getName());
+			}
+		});
+
+		for (Resource child : childs) {
+			serializeTreeNewMatches(child, queries, buffer);
+		}
+	}
+
+	/**
+	 * Serialize method from LinkFormat class with old method
+	 * 
+	 * @param resource
+	 * @param queries
+	 * @param buffer
+	 */
+	public static void serializeTreeOldMatches(Resource resource, List<String> queries, StringBuilder buffer) {
+		// add the current resource to the buffer
+		if (resource.isVisible() && matchesOld(resource, queries)) {
+			buffer.append(LinkFormat.serializeResource(resource));
+		}
+
+		// sort by resource name
+		List<Resource> childs = new ArrayList<Resource>(resource.getChildren());
+		Collections.sort(childs, new Comparator<Resource>() {
+			@Override
+			public int compare(Resource o1, Resource o2) {
+				return o1.getName().compareTo(o2.getName());
+			}
+		});
+
+		for (Resource child : childs) {
+			serializeTreeOldMatches(child, queries, buffer);
+		}
+	}
+
+	/**
+	 * Matches method before fix
+	 * 
+	 * @param resource
+	 * @param queries
+	 * @return
+	 */
+	public static boolean matchesOld(Resource resource, List<String> queries) {
+
```

Comment:
Again, I might have confused you but I didn't mean to reproduce the actual matcher implementation here in the test case. In particular not the old (erroneous) one...